### DISCUSSION TITLE
How to overload an operator ?

### DISCUSSION THREAD

**Discussion by srilakshmikanthanp** [josdejong/mathjs#2937] [Q&A] [answered]
Can I overload an operator like `+` to evaluate custom types like,

~~~js
const add = (one, two) => {
 if (typeof one === 'object' && typeof two === 'number') {
    for (let key in obj) {
      obj[key] += two;
    }
 }

 // handle other cases ...
}

const getObject = () => {
  return { one: 2, two: 2 }
}

evaluate(`getObject() + 2') // { one: 4, two: 4 }
~~~

How do I put the overloaded add to the evaluator?

**Comment by srilakshmikanthanp** [comment]
Or Can I implement a custom datatype to do the tasks, how to do it?

**Reply by dvd101x** [reply, marked as answer]
You could import to your function a typed function with a new type signature.
```js
math.import(
    {
        add: math.typed(
            {
                'Object, number': (obj, num) => {
                    for (key in obj){
                        obj[key] += num
                    }
                    return obj
                }
            }
        )
    }
)
```
## Reference
https://mathjs.org/docs/core/extension.html

**Reply by dvd101x** [reply]
It is also possible to add new custom types and add convertions between types.

But if the type already exists (like object) then the best way is to import the missing type signatures to the function you need.

## Reference
https://mathjs.org/docs/core/extension.html#typed-functions
https://github.com/josdejong/typed-function
